pkg/kubelet/kuberuntime/container.go

```diff
@@ -0,0 +1,440 @@
+	if err != nil {
+		return msg, err
+	}
+
+	ref, err := kubecontainer.GenerateContainerRef(pod, container)
+	if err != nil {
+		glog.Errorf("Can't make a ref to pod %q, container %v: '%v'", format.Pod(pod), container.Name, err)
+	}
+	glog.V(4).Infof("Generating ref for container %s: %#v", container.Name, ref)
+
+	// For a new container, the RestartCount should be 0
+	restartCount := 0
+	containerStatus := podStatus.FindContainerStatusByName(container.Name)
+	if containerStatus != nil {
+		restartCount = containerStatus.RestartCount + 1
+	}
+
+	containerConfig, err := m.generateContainerConfig(container, pod, restartCount)
+	if err != nil {
+		m.recorder.Eventf(ref, api.EventTypeWarning, events.FailedToCreateContainer, "Failed to create container with error: %v", err)
+		return "Generate Container Config Failed", err
+	}
+
+	containerID, err := m.runtimeService.CreateContainer(podSandboxID, containerConfig, podSandboxConfig)
+	if err != nil {
+		m.recorder.Eventf(ref, api.EventTypeWarning, events.FailedToCreateContainer, "Failed to create container with error: %v", err)
+		return "Create Container Failed", err
+	}
+	m.recorder.Eventf(ref, api.EventTypeNormal, events.CreatedContainer, "Created container with id %v", containerID)
+
+	if ref != nil {
+		m.containerRefManager.SetRef(kubecontainer.ContainerID{
+			Type: m.runtimeName,
+			ID:   containerID,
+		}, ref)
+	}
+
+	err = m.runtimeService.StartContainer(containerID)
+	if err != nil {
+		m.recorder.Eventf(ref, api.EventTypeWarning, events.FailedToStartContainer,
+			"Failed to start container with id %v with error: %v", containerID, err)
+		return "Start Container Failed", err
+	}
+	m.recorder.Eventf(ref, api.EventTypeNormal, events.StartedContainer, "Started container with id %v", containerID)
+
+	if container.Lifecycle != nil && container.Lifecycle.PostStart != nil {
+		kubeContainerID := kubecontainer.ContainerID{
+			Type: m.runtimeName,
+			ID:   containerID,
+		}
+		msg, handlerErr := m.runner.Run(kubeContainerID, pod, container, container.Lifecycle.PostStart)
+		if handlerErr != nil {
+			err := fmt.Errorf("PostStart handler: %v", handlerErr)
+			m.generateFailedContainerEvent(kubeContainerID, pod.Name, events.FailedPostStartHook, msg)
+			m.killContainer(pod, kubeContainerID, container, "FailedPostStartHook", nil)
+			return "PostStart Hook Failed", err
+		}
+	}
+
+	return "", nil
+}
+
+// makeMounts generates container volume mounts for kubelet runtime api.
+func makeMounts(containerFullName string, opts *kubecontainer.RunContainerOptions, container *api.Container) []*runtimeApi.Mount {
+	volumeMounts := []*runtimeApi.Mount{}
+
+	for _, v := range opts.Mounts {
+		volumeMounts = append(volumeMounts, &runtimeApi.Mount{
+			Name:           &v.Name,
+			HostPath:       &v.HostPath,
+			ContainerPath:  &v.ContainerPath,
+			Readonly:       &v.ReadOnly,
+			SelinuxRelabel: &v.SELinuxRelabel,
+		})
+	}
+
+	if opts.PodContainerDir != "" && len(container.TerminationMessagePath) != 0 {
```

Comment:
Could you also copy the original comments if they still apply here? E.g.,
https://github.com/kubernetes/kubernetes/blob/007257a5adac832d276af10cbe0c617e6189f04a/pkg/kubelet/dockertools/docker_manager.go#L611
https://github.com/kubernetes/kubernetes/blob/007257a5adac832d276af10cbe0c617e6189f04a/pkg/kubelet/dockertools/docker_manager.go#L618